javaslang/src/main/java/javaslang/concurrent/Future.java

```diff
@@ -247,6 +249,38 @@
     }
 
     /**
+     * Creates a {@code Future} with the given java.util.concurrent.CompletableFuture, backed by the {@link #DEFAULT_EXECUTOR_SERVICE}
+     *
+     * @param future A {@link java.util.concurrent.CompletableFuture}
+     * @param <T>    Result type of the Future
+     * @return A new {@code Future} wrapping the result of the {@link java.util.concurrent.CompletableFuture}
+     * @throws NullPointerException if future is null
+     */
+    static <T> Future<T> fromCompletableFuture(CompletableFuture<T> future) {
+        return fromCompletableFuture(DEFAULT_EXECUTOR_SERVICE, future);
+    }
+
+    /**
+     * Creates a {@code Future} with the given java.util.concurrent.CompletableFuture, backed by given {@link ExecutorService}
+     *
+     * @param executorService An {@link ExecutorService}
+     * @param future          A {@link java.util.concurrent.CompletableFuture}
+     * @param <T>             Result type of the Future
+     * @return A new {@code Future} wrapping the result of the {@link java.util.concurrent.CompletableFuture}
+     * @throws NullPointerException if executorService or future is null
+     */
+    static <T> Future<T> fromCompletableFuture(ExecutorService executorService, CompletableFuture<T> future) {
+        Objects.requireNonNull(executorService, "executorService is null");
+        Objects.requireNonNull(future, "future is null");
+        FutureImpl<T> newFuture = new FutureImpl<>(executorService);
+        future.handle((t, throwable) -> newFuture
+                .tryComplete(throwable == null
+                             ? Try.success(t)
+                             : Try.failure(throwable instanceof CompletionException ? throwable.getCause() : throwable)));
+        return newFuture;
```

Comment:
The FutureImpl is isolated - there are exactly two occurences/usages: In Future.of() and in Promise.make(). All other API should build on these and not use FutureImpl directly. This scales better from the design perspective.

In our case we should use the following impl:

```java
final Promise<T> promise = Promise.make();
future.handle((t, err) -> {
    if (err == null) {
        promise.success(t);
    } else {
        promise.failure(err instanceof CompletionException ? err.getCause() : err);
    }
});
return promise.future();
```

**Note:** The upcoming release will (but currently does not) contain a more concise syntax:

```java
return Future.of(p -> completableFuture.handle((t, err) -> {
    if (err == null) {
        p.success(t);
    } else {
        p.failure(err instanceof CompletionException ? err.getCause() : err);
    }
});
```